File: c/problem_45.c
#include "problem_45.h"
#include <stdio.h>
#include <stdlib.h>
#include <assert.h>

static int rand5(void) { return rand() % 5 + 1; }

int rand7_p45(void) {
    while (1) {
        int n = (rand5()-1)*5 + (rand5()-1);
        if (n < 21) return n % 7 + 1;
    }
}

void problem45(void) {
    printf("Problem 45:\n");
    int counts[8] = {0};
    for (int i = 0; i < 70000; i++) {
        int v = rand7_p45();
        assert(v >= 1 && v <= 7);
        counts[v]++;
    }
    for (int i = 1; i <= 7; i++)
        assert(counts[i] > 5000 && counts[i] < 15000);
    printf("All Tests Passed\n");
}

File: typescript/problem_45.ts
function rand5ts(): number { return Math.floor(Math.random() * 5) + 1; }

export function rand7ts(): number {
    while (true) {
        const n = (rand5ts() - 1) * 5 + (rand5ts() - 1);
        if (n < 21) return n % 7 + 1;
    }
}

export function problem45(): void {
    console.log("Problem 45:");
    const counts = new Array(8).fill(0);
    for (let i = 0; i < 70000; i++) {
        const v = rand7ts();
        if (v < 1 || v > 7) throw new Error("out of range");
        counts[v]++;
    }
    for (let i = 1; i <= 7; i++)
        if (counts[i] < 7000 || counts[i] > 13000) throw new Error("bad distribution");
    console.log("All Tests Passed");
}
